Adding Spatter

mahyarsamani opened the pull request:
This PR adds source code for C++ implementation of SpatterGen as well as SpatterKernel. SpatterGen uses a PyBindMethod to add kernels to the backend code. This way the process of processing json files could be offloaded to python. In addition it adds standard library components for SpatterGenCore and SpatterGen. These two components follow the same structure as AbstractCore and AbstractProcessor. In addition spatter_kernel.py adds a definition for SpatterKernel in python to make adding kernels to C++ easier. Also it adds utility functions for parsing dictionaries read from json as well as partitioning traces for multicore setups.

mahyarsamani committed:
cpu: Adding SpatterGen

This change adds source code for SpatterGen ClockedObject.
The set of source code pushed includes code for SpatterKernel
that tracks whether information is being gathered or scattered
as well as the list of indices to be accessed. This model
has PyBindMethod to add SpatterKernels from python.
This way all the preparations for kernels can be done in python.
SpatterGen has a few parameters that model limits on a few of
hardware resources in the backend of a processor, e.g. number
of functional units to calculate effective address, the latency
of calculating effective address, number of integer registers.

Change-Id: I451ffb385180a914e884cab220928c5f1944b2e3

powerjg commented:
I know this is a draft, but a few quick comments:
1. You need to add copyright blocks
2. The names should be something more descriptive than "model"
3. The commit message and the PR description need to have details

mahyarsamani commented on src/base/stats/units.hh:
```diff
@@ -174,6 +174,23 @@ class Second : public Base
     std::string getUnitString() const override { return Second::toString(); }
 };
 
+class NanoSecond : public Base
```
@powerjg If I want to express latency stats in seconds, due to precision limitations they will all be zero. I added this new unit to express the latencies in nanoseconds. I am not really sure if this is what the right way to do this is. Let me know if I have to change this to Tick.

powerjg replied:
Yeah, I'm against this. Just use `Ticks` if you need high precision.

mahyarsamani committed:
stdlib: Adding SpatterGenCore and SpatterGen

This change adds code for SpatterGenCore and SpatterGen as well
as SpatterKernel to the standard library. SpatterGenCore and
SpatterGen follow the same structure as AbstractCore and
AbstractProcessor. spatter_kernel.py adds utility functions
to parse dictionaries as well as partition a list into
multiple lists through interleaving to be used when setting up
a multicore SpatterGen.

Change-Id: I003553e97f901c0724f5feac0bb6e21a020bd6ad

powerjg commented on src/cpu/testers/spatter_gen/SpatterGen.py:
```diff
@@ -0,0 +1,55 @@
+    request_buffer_size = Param.Int("Size of the request buffer.")
+    send_bandwidth = Param.Int("Number of requests to send in parallel.")
+
+    cxx_exports = [PyBindMethod("addKernel")]
```
Can you add a reference to the spatter paper for this class? Call `add_citation(SpatterGen, <bibtex text>)` at the end of this file.

powerjg commented on src/cpu/testers/spatter_gen/SpatterGen.py:
```diff
@@ -0,0 +1,55 @@
+
+    int_regfile_size = Param.Int("Size of the integer register file.")
+    fp_regfile_size = Param.Int("Size of the floating point register file.")
+    request_gen_latency = Param.Int(
```
Should this be a `Param.Cycles` if it's cycles?

powerjg commented on src/cpu/testers/spatter_gen/SpatterGen.py:
```diff
@@ -0,0 +1,55 @@
+    request_gen_latency = Param.Int(
+        "Number of cycles to spend for creating a request."
+    )
+    request_gen_bandwidth = Param.Int("Number of requests generate per cycle.")
```
```suggestion
    request_gen_rate = Param.Int("Number of requests generate per cycle.")
```
Bandwidth isn't exactly the best term here.

powerjg commented on src/cpu/testers/spatter_gen/SpatterGen.py:
```diff
@@ -0,0 +1,55 @@
+        "Number of cycles to spend for creating a request."
+    )
+    request_gen_bandwidth = Param.Int("Number of requests generate per cycle.")
+    request_buffer_size = Param.Int("Size of the request buffer.")
```
```suggestion
    request_buffer_entries = Param.Int("Size of the request buffer.")
```
To me, `size` implies it is in `Bytes`

powerjg commented on src/cpu/testers/spatter_gen/SpatterGen.py:
```diff
@@ -0,0 +1,55 @@
+    )
+    request_gen_bandwidth = Param.Int("Number of requests generate per cycle.")
+    request_buffer_size = Param.Int("Size of the request buffer.")
+    send_bandwidth = Param.Int("Number of requests to send in parallel.")
```
I don't understand how this is different from `request_gen_bandwidth`. The name and/or help text needs to be updated to make it more clear.

powerjg commented on src/cpu/testers/spatter_gen/enums.cc:
```diff
@@ -0,0 +1,44 @@
+    "ASLEEP"
+};
+
+} // namespace gem5
```
You shouldn't need this file at all. See the code here for how it works: https://github.com/gem5/gem5/blob/stable/src/base/stl_helpers/ostream_helpers.hh#L81

powerjg commented on src/cpu/testers/spatter_gen/spatter_gen.hh:
```diff
@@ -0,0 +1,227 @@
+{
+  private:
+
+    class SleepyEvent : public EventFunctionWrapper
```
Can you use a less generic name, like "SpatterGenEvent"? I think I see how you want this to be generic, but I don't think it makes sense to make it generic here and now 🙂

powerjg commented on src/cpu/testers/spatter_gen/spatter_gen.hh:
```diff
@@ -0,0 +1,227 @@
+    {
+      // TODO: split pending into pendingInput and pendingOutput
+      private:
+        SleepyState _state;
```
```suggestion
        SpatterGen::State _state;
```

I would make this private to the `SpatterGen` class.

powerjg replied:
Don't forget about this.

powerjg commented on src/cpu/testers/spatter_gen/enums.hh:
```diff
@@ -0,0 +1,53 @@
+    ASLEEP,
+    NUM_SLEEPY_STATE
+};
+extern const char* sleepyStateStrings[NUM_SLEEPY_STATE];
```
I would make this private to the `SpatterGen` class

powerjg commented on src/cpu/testers/spatter_gen/spatter_gen.hh:
```diff
@@ -0,0 +1,227 @@
+    int requestBufferSize;
+    std::queue<PacketPtr> requestBuffer;
+    // same explanation as receiveBufferInsertionTime
+    std::queue<Tick> requestBufferInsertionTime;
```
I don't understand this comment

powerjg commented on src/cpu/testers/spatter_gen/spatter_gen.hh:
```diff
@@ -0,0 +1,227 @@
+    // we will assume it takes 1 cycle to issue memory requests
+    int sendBandwidth;
+    Tick firstPortAvailableTime;
+    std::vector<Tick> portBusyUntil;
```
Why is this a vector?

mahyarsamani replied:
I need to access it with index.

powerjg commented on src/cpu/testers/spatter_gen/spatter_gen.hh:
```diff
@@ -0,0 +1,227 @@
+    // no need to limit the size of this buffer.
+    // it's a response buffer and it will automatically
+    // be limited by requestBufferSize, intRegFileSize, fpRegFileSize
+    std::queue<PacketPtr> receiveBuffer;
```
I'm concerned these are queues. Do we have guarantees on ordering?

It also seems like there are pairs of `PacketPtr` and `Tick`. Why not use one datastructure?

mahyarsamani replied:
Answer to second question is, it's a preference.
About your first question, do you mean that responses do not arrive at the same order requests were issued? If so, I can ensure that all responses before an expected response have arrived before another access is made based on the resp.

powerjg commented on src/python/gem5/components/processors/spatter_generator.py:
```diff
@@ -0,0 +1,103 @@
+                clock=clk_freq, voltage_domain=VoltageDomain()
+            )
+            for generator in self.cores:
+                generator.clk_domain = clock_domain
```
It would be good to add a comment here about why you don't need an `else` statement.

powerjg commented on src/python/gem5/components/processors/spatter_generator_core.py:
```diff
@@ -0,0 +1,70 @@
+
+    def start_traffic(self) -> None:
+        for kernel in self._kernels:
+            print(f"Adding kernel: {kernel}")
```
Probably should drop this print for upstream.

powerjg commented on src/python/gem5/components/processors/spatter_kernel.py:
```diff
@@ -0,0 +1,103 @@
+    return partitions
+
+
+class SpatterKernel:
```
This class should really come with some help comments explaining how it's used. Pointers to spatter documentation is great.

powerjg commented on src/python/gem5/components/processors/spatter_generator.py:
```diff
@@ -0,0 +1,103 @@
+# Copyright (c) 2024 The Regents of the University of California
```
I would put these files in a subdirectory. You can expose them in the `processors` module via `__init__.py` if you want, but this directory is getting out of hand.

powerjg commented on src/cpu/testers/spatter_gen/spatter_gen.cc:
```diff
@@ -0,0 +1,584 @@
+    bool no_queued_req = requestBuffer.empty();
+    bool no_index = receiveBuffer.empty();
+    bool no_kernel = kernels.empty();
+    if (no_pending_mem && no_queued_req && no_index && no_kernel)
+    {
+        exitSimLoop(
+            csprintf("SpatterGen %s done generating requests.", name()),
+            0,
+            nextCycle()
+        );
+    }
```
Is this how the other generators work? I thought if they had nothing to do, they would just not schedule any events and you would get a `MaxTick` exit. I would like all generators to have the same behavior.

powerjg replied:
What do you think about this?

mahyarsamani replied:
You are right, this behavior is different. However, I remember having much trouble with using bandwidth stats with that behavior.

powerjg commented on src/cpu/testers/spatter_gen/spatter_gen.cc:
```diff
@@ -0,0 +1,584 @@
+    }
+}
+
+void
+SpatterGen::SpatterGenPort::sendPacket(PacketPtr pkt) {
+    panic_if(blocked(), "Should never try to send if port is blocked.");
```
```suggestion
void
SpatterGen::SpatterGenPort::sendPacket(PacketPtr pkt)
{
    panic_if(blocked(), "Should never try to send if port is blocked.");
```

powerjg commented on src/cpu/testers/spatter_gen/spatter_gen.cc:
```diff
@@ -0,0 +1,584 @@
+        // TODO: record travel time for indexer accesses.
+        receiveBuffer.push(pkt);
+        receiveBufferInsertionTime.push(curTick());
+        // REMEMBER: might need to wake nextGenEvent up?
+        // IDEA: Separate pending to pendingInput and pendingOutput?
+        // I think for now we should not wake up nextGenEvent,
+        // because nextGenEvent does not sleep if there are not
+        // enough registers available, which is something that
+        // this scope of the if does not influence.
```
Is this a future todo?

mahyarsamani replied:
Yes

powerjg commented on src/cpu/testers/spatter_gen/spatter_gen.cc:
```diff
@@ -0,0 +1,584 @@
+        } else {
+            panic("Unknown access type.");
+        }
+        // CAUTION: We're going to decrement fpRegUsed here,
+        // it could cause inaccuracies if processNextGenEvent
+        // is called after recvTimingResp.
+        // it may be fine though since we don't simulate the rest,
+        // it's probably not going to ever be an issue since
+        // fpRegFileSize is probably >> requestBufferSize
```
Would it make sense to put a panic/assert here?

mahyarsamani replied:
I will put a panic telling user to make fpRegFileSize bigger than request_buffer_entries.

powerjg commented on src/cpu/testers/spatter_gen/spatter_gen.cc:
```diff
@@ -0,0 +1,584 @@
+
+            front.iterate();
+            if (front.done()) {
+                exitSimLoop(
+                    csprintf(
+                    "SpatterGen %s done generating "
+                    "requests from SpatterKernel %d.",
+                    name(), front.id()
+                    ),
+                    0,
+                    nextCycle()
+                );
```
Why is this not done in `checkSimDone`?

mahyarsamani replied:
checkSimDone is for overall exit. This exit is per kernel. We're not completely done here. I think it is useful to try dump stats per kernel.

powerjg commented on src/cpu/testers/spatter_gen/spatter_gen.cc:
```diff
@@ -0,0 +1,584 @@
+            int_reg_used_delta--;
+            receiveBuffer.pop();
+            receiveBufferInsertionTime.pop();
+        } else if (have_kernel && have_int_reg) {
```
I would split these two `if`/`else if` clauses into their own functions. This function is too big.

mahyarsamani replied:
I agree. I would argue for not changing it right now. I will update this after we have a more definitive picture of what we want the SpatterKernel class to look like.

powerjg replied:
I would argue that if we don't know what we want the kernel to look like, then it's not ready to be merged upstream. We should go ahead and do it the right way the first time. Once something is merged, the activation energy to change it is very high.

powerjg reviewed (commented):
Overall, this looks great. I didn't check the logic carefully in `spatter_gen.cc`, but I trust you've tested this and compared with upstream spatter.

Most of my comments are style. The only "big" thing that I would like to see changed is moving the state information into the `SpatterGen` class to make it private. 

One thing to keep in mind when pushing code upstream is that you should try to make things as self-contained as possible. Don't expose things publicly unless absolutely required.

Finally, there are some places where you need to add more documentation. Think about how someone would discover how to use this if they don't know anything about Spatter and they are not involved in our project.

mahyarsamani commented:
@powerjg @giactra 
Still working on addressing some comments. Will fix by tomorrow.

powerjg commented on src/python/gem5/components/processors/spatter_gen/__init__.py:
```diff
@@ -0,0 +1,6 @@
+from .spatter_generator import SpatterGenerator
```
Copyright...

powerjg commented on src/cpu/testers/spatter_gen/spatter_gen.cc:
```diff
@@ -0,0 +1,595 @@
+            "Ideally: fp_regfile_size >> request_buffer_entries."
+    );
+    generatorBusyUntil.resize(requestGenRate);
+    for (int i = 0; i < requestGenRate; i++) {
+        generatorBusyUntil[i] = 0;
+    }
```
You don't need these for loops. You can set the value to use when calling `resize`. See https://en.cppreference.com/w/cpp/container/vector/resize

powerjg commented on src/python/gem5/components/processors/spatter_gen/spatter_kernel.py:
```diff
@@ -0,0 +1,181 @@
+    """This class encapsulates one kernel in a spatter trace.
+    A spatter trace is represented with a json file.
+    An example of a spatter trace can be found here:
+        bit.ly/3X2IijM
```
Can you point to the repo instead? I don't trust that bitly will be around in 20 years.

powerjg commented on src/cpu/testers/spatter_gen/spatter_gen.cc:
```diff
@@ -0,0 +1,591 @@
+    firstPortAvailableTime(0),
+    nextSendEvent([this](){ processNextSendEvent(); }, name() + ".SendEvent")
+{
+    panic_if(fpRegFileSize < requestBufferEntries,
```
`fatal`? If this is a user configuration error it should be fatal. Panic is for bugs (like an assert)

mahyarsamani replied:
Will do

powerjg commented on src/cpu/testers/spatter_gen/spatter_gen.cc:
```diff
@@ -0,0 +1,591 @@
+void
+SpatterGen::SpatterGenPort::recvReqRetry()
+{
+    DPRINTF(SpatterGen, "%s: Port received a ReqRetry.\n", __func__);
```
I would add an `assert(blockedPacket);` here

mahyarsamani replied:
Will do

powerjg commented on src/cpu/testers/spatter_gen/spatter_gen.hh:
```diff
@@ -0,0 +1,239 @@
+namespace gem5
+{
+
+typedef std::tuple<PacketPtr, Tick> TimedPacket;
+
+PacketPtr entryPacket(const TimedPacket& timed_pkt);
+Tick entryTick(const TimedPacket& timed_pkt);
```
1. Where are these defined?
2. Why are they not in the `SpatterGen` class? Do you have any uses `TimedPacket` outside of the class members?

mahyarsamani replied:
They are defined in spatter_gen.cc. It's a matter of preference for me to not define it in the SpatterGen class. I can move it inside if there is a right/wrong.

powerjg replied:
They are polluting the gem5 namespace is the problem. I would put them in the class until they are used outside of it.

mahyarsamani replied:
I see, will do.

powerjg commented on src/cpu/testers/spatter_gen/spatter_gen.hh:
```diff
@@ -0,0 +1,239 @@
+    {
+      // TODO: split pending into pendingInput and pendingOutput
+      private:
+        enum SleepState
+        {
+            AWAKE,
+            ASLEEP
+        };
+
+        SleepState _state;
```
Seems like this could just be a boolean `isPending` (or `isWaiting` or something like that)

mahyarsamani replied:
I like to keep it an enum, I know right now the state is a boolean like value. In future I want to extend this so that it can encode more states, e.g. ASLEEP_ON_INPUT, ASLEEP_ON_OUTPUT, ...

powerjg commented on src/cpu/testers/spatter_gen/spatter_gen.hh:
```diff
@@ -0,0 +1,239 @@
+            EventFunctionWrapper(callback, name), _state(SleepState::AWAKE)
+        {}
+        // a SpatterGenEvent will only be asleep if it is pending output
+        bool pending() { return _state == SleepState::ASLEEP; }
+        void sleep() { _state = SleepState::ASLEEP; }
```
What is the difference between `pending()` and `sleep()`?

mahyarsamani replied:
pending is a getter and sleep is a setter.

powerjg reviewed (commented):
A few more little things

powerjg reviewed (dismissed):
Looks good to me. Thanks for the updates!

mahyarsamani commented:
@giactra This is ready for review.

powerjg reviewed (changes requested):
There's a merge/rebase issue here.

mahyarsamani commented:
> There's a merge/rebase issue here.

I think it should be fixed now.

powerjg commented:
@giactra let me know if you want us to hold off merging until you have time to review.

giactra commented on src/cpu/testers/spatter_gen/spatter_gen.hh:
```diff
@@ -0,0 +1,224 @@
+using enums::SpatterKernelType;
+using enums::SpatterProcessingMode;
+
+class SpatterGen: public ClockedObject
```
I appreciate there is a reference to the publication in the SpatterGen.py file, but I believe it would be beneficial for a user to have a brief description of Spatter classes and what they are actually doing, so that a user is not "forced" to read the paper, especially if the paper requires a subscription

mahyarsamani replied:
Would the documentation in spatter_kernel.py suffice? I am now looking at writing it up and wonder if I should extend what is already there.

giactra commented on src/cpu/testers/spatter_gen/spatter_gen.hh:
```diff
@@ -0,0 +1,224 @@
+namespace gem5
+{
+
+using enums::SpatterKernelType;
+using enums::SpatterProcessingMode;
```
Is there a reason to make them available like this in the gem5 namespace? Shall we restrict the visibility to the Spatter class?

giactra commented on src/cpu/testers/spatter_gen/spatter_gen.hh:
```diff
@@ -0,0 +1,224 @@
+    {
+      private:
+        // TODO: split pending into pendingInput and pendingOutput
+        enum SleepState
```
I would use scoped enum here

giactra commented on src/cpu/testers/spatter_gen/spatter_gen.hh:
```diff
@@ -0,0 +1,224 @@
+            EventFunctionWrapper(callback, name), _state(SleepState::AWAKE)
+        {}
+        // a SpatterGenEvent will only be asleep if it is pending output
+        bool pending() { return _state == SleepState::ASLEEP; }
```
Nitpick: this could be marked as const

giactra commented on src/cpu/testers/spatter_gen/spatter_gen.hh:
```diff
@@ -0,0 +1,224 @@
+        SpatterGenStats(SpatterGen* spatter_gen);
+    };
+
+    enum SpatterGenState
```
Ditto

giactra commented on src/cpu/testers/spatter_gen/utility_structs.hh:
```diff
@@ -0,0 +1,240 @@
+namespace gem5
+{
+
+using enums::SpatterKernelType;
```
Ditto

giactra commented on src/cpu/testers/spatter_gen/utility_structs.hh:
```diff
@@ -0,0 +1,240 @@
+    }
+};
+
+typedef std::tuple<Addr, size_t> AccessPair;
```
I would move this within the SpatterAccess class

giactra commented on src/cpu/testers/spatter_gen/utility_structs.hh:
```diff
@@ -0,0 +1,240 @@
+*/
+
+#ifndef __CPU_TESTERS_SPATTER_GEN_SPATTER_KERNEL_HH__
+#define __CPU_TESTERS_SPATTER_GEN_SPATTER_KERNEL_HH__
```
This is not matching the header file

giactra commented on src/cpu/testers/spatter_gen/utility_structs.hh:
```diff
@@ -0,0 +1,240 @@
+        return item;
+    }
+
+    T front() { return items.front(); }
+
+    bool empty() { return items.empty(); }
+
+    size_t size() { return items.size(); }
```
Could be marked as const

giactra commented on src/cpu/testers/spatter_gen/utility_structs.hh:
```diff
@@ -0,0 +1,240 @@
+    SpatterAccess(
+        RequestorID requestor_id,
+        SpatterKernelType kernel_type,
+        std::queue<AccessPair> access_pairs
```
I would pass by const reference here as you are making a copy

giactra commented on src/cpu/testers/spatter_gen/utility_structs.hh:
```diff
@@ -0,0 +1,240 @@
+        accTripTime(0), accessPairs(access_pairs)
+    {}
+
+    SpatterKernelType type() { return _kernelType; }
+
+    int tripsLeft() { return accessPairs.size(); }
```
could be const

giactra commented on src/cpu/testers/spatter_gen/utility_structs.hh:
```diff
@@ -0,0 +1,240 @@
+
+    void recordTripTime(Tick trip_time) { accTripTime += trip_time; }
+
+    Tick tripTimeSoFar() { return accTripTime; }
```
could be const

giactra commented on src/cpu/testers/spatter_gen/utility_structs.hh:
```diff
@@ -0,0 +1,240 @@
+        remRolls = indices.size();
+    }
+
+    SpatterKernelType type() { return _type; }
+
+    bool done() { return iteration == count; }
```
Could be const

giactra commented on src/cpu/testers/spatter_gen/utility_structs.hh:
```diff
@@ -0,0 +1,240 @@
+        index(0), iteration(0), remRolls(0)
+    {}
+
+    uint32_t id() { return _id; }
```
Could be const

giactra reviewed (dismissed):
Thanks @powerjg, I am not going to do a full functionality check assuming it has been tested and works properly already. I only wanted to make sure the code was properly documented (that's the main comment I have if you see below). Also some minor comments on the code from a SW engineering point of view

BobbyRBruce commented:
@powerjg : Can you sign off on this so we can get it in for the v24.0 staging?
